ddd/ddd-05-mem-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,250 @@
+  not concatenatable, whereas TCP uses a byte stream model. So existing
+  applications based around this approach will assume they can pull data out
+  of the memory BIOs in arbitrary-length chunks and feed them to the
+  network, which is not the case. Moreover, if libssl just writes the UDP
+  datagram payload to the memory BIO for each UDP datagram it wants ton
+  send, the BIO will end up containing the concatenation of these datagrams
+  with no way for the application to determine where one datagram should
+  start and end.
+
+  - One possible solution is to put a length header on each UDP datagram.
+    However this requires the application to parse this length header and
+    parse datagrams out of the BIO bytestream. Since datagrams will probably
+    not happen to be aligned at the start of the application's call to
+    BIO_read, this will require the application to copy partial datagrams to
+    a staging area in order to "unfragment" them before finally sending
+    them. This is neither performant nor a reasonable ask upon the
+    application.
+
+  - Looking at the big picture, the *reason* an application wants to use
+    a memory BIO on the network side is so it can use libssl as a pure
+    state machine and be in control of all I/O.
+
+    Since supporting the exact same interface (memory BIOs with bytestream
+    semantics) is infeasible for the reasons given above, the next best thing we
+    can do is to ensure that we accommodate this use case for applications, so
+    that there is some rougly analagous solution available to them. This may not
+    be "minimal changes" in the sense of "the changes are small in an absolute
+    sense" but it does meet the definition of "minimal changes" where "minimal"
+    means "as small as actually possible". While adaptation of this kind of
+    application (demo 5) will require more work than the other kinds of
+    application (as demonstrated in demos 1 through 4) we can at least provide
+    an interface which applications will find familiar and which lacks any
+    excessive impedence discontinuities with their own code.
+
+  - There already exists BIO_dgram, which is used with DTLS. This causes
+    BIO_read/BIO_write to work similarly to recv/send. This works on
+    a file descriptor; there is no analagous BIO_dgram_mem.
+    Such a BIO could be created, which would model datagram semantics
+    for read/write operations (i.e., a sequence of BIO_write calls
+    with given lengths results in an exactly matching sequence of
+    BIO_read calls with the same lengths returned).
+
+  - There are some caveats for this idea. In order for this to work:
+
+    - applications must ensure there is a guaranteed 1:1 correspondance
+      between send/recv and BIO_read/BIO_write calls; and
+
+    - the buffer passed to BIO_read() and recv() must always be big
+      enough to receive the largest possible packet, else silent
+      truncation occurs.
```

Comment:
We can't do much to help applications in calling`recv` this is outside of OpenSSL. For a `BIO_dgram_mem` we can fail if the buffer passed to `BIO_read()` isn't big enough. We can also supply a new BIO ctrl to find out the size of the next packet in the mem buffer.

Reply:
> I like this idea, but this causes the semantics of a BIO_dgram (which is just a wrapper around recvfrom) to differ from those of BIO_dgram_mem, so it doesn't really seem like an option. The entire point is for these to have the same semantics.

I wonder how much this matters? It means that BIO_dgram_mem is slightly stricter that BIO_dgram - but it is stricter in the case where BIO_dgram will silently fail which is probably unwanted behaviour.